### PR TITLE
Feature/block hash to number

### PR DESCRIPTION
**Pull request by cshcomcom**
## Description
- block-to-hash in memory -> block-to-hash in disk

### CODE REVIEW

File: blockchain/block-file-util.js [hunk 1]
```diff
@@ -3,32 +3,37 @@ const glob = require('glob');
...
-  static getAllBlockFiles(chainPath) {
-    const allBlockFilesPattern = `${chainPath}/*.${FILE_NAME_SUFFIX}`;
+  static getAllBlockPaths(chainPath) {
+    const allBlockFilesPattern = `${chainPath}/${BLOCKCHAINS_N2B_DIR_NAME}/*.${FILE_NAME_SUFFIX}`;
     return glob.sync(allBlockFilesPattern).sort(compare());
   }
 
-  static getBlockFiles(chainPath, from, to) {
+  static getBlockPaths(chainPath, from, to) {
     const blockFiles = [];
```
Comment (liayoo) [inline blockchain/block-file-util.js:34]: blockPaths?

File: common/constants.js [hunk 2]
```diff
@@ -22,7 +22,9 @@ if (!fs.existsSync(PROTOCOL_VERSIONS)) {
   throw Error('Missing protocol versions file: ' + PROTOCOL_VERSIONS);
 }
 const PROTOCOL_VERSION_MAP = JSON.parse(fs.readFileSync(PROTOCOL_VERSIONS));
-const BLOCKCHAINS_DIR = path.resolve(__dirname, '../blockchain/blockchains');
+const BLOCKCHAINS_DIR = path.resolve(__dirname, '../blockchains');
+const BLOCKCHAINS_N2B_DIR_NAME = 'n2b';
```
Comment (liayoo) [inline common/constants.js:26]: Could you leave some notes about the names (n2b, h2n)?

File: blockchain/block-file-util.js [hunk 3]
```diff
@@ -37,41 +42,54 @@ class BlockFileUtil {
...
-    let created = false;
-    const dirs = [BLOCKCHAINS_DIR];
-    if (chainPath) {
-      dirs.push(chainPath);
+    const n2bPath = path.join(chainPath, BLOCKCHAINS_N2B_DIR_NAME);
+    const h2nPath = path.join(chainPath, BLOCKCHAINS_H2N_DIR_NAME);
+
+    if (!fs.existsSync(chainPath)) {
+      fs.mkdirSync(chainPath, {recursive: true});
+      fs.mkdirSync(n2bPath);
```
Comment (platfowner) [inline blockchain/block-file-util.js:50]: It seems it's not handling the cases that chainPath exists without n2bPath or h2nPath?

File: blockchain/block-file-util.js [hunk 4]
```diff
@@ -37,41 +42,54 @@ class BlockFileUtil {
...
+  static writeBlock(chainPath, block) {
+    const blockPath = this.getBlockPath(chainPath, block.number);
+    if (!fs.existsSync(blockPath)) {
+      const compressed = zlib.gzipSync(Buffer.from(JSON.stringify(block)));
+      fs.writeFileSync(blockPath, compressed);
+    }
+  }
+
+  static writeHashToNumber(chainPath, blockHash, blockNumber) {
+    if (!blockHash || (blockNumber !== 0 && !blockNumber)) {
```
Comment (platfowner) [inline blockchain/block-file-util.js:80]: It looks like `blockNumber !== 0 && !blockNumber` is identical to `!blockNumber`?

And isn't the genesis block's number 0?

What about `!blockHash || !ChainUtil.isNumber(blockNumber)`?

File: blockchain/block-file-util.js [hunk 5]
```diff
@@ -37,41 +42,54 @@ class BlockFileUtil {
...
+    return this.readBlock(blockPath);
   }
 
   // TODO(csh): Change to asynchronous
-  static writeBlock(filePath, block) {
-    const compressed = zlib.gzipSync(Buffer.from(JSON.stringify(block)));
-    fs.writeFileSync(filePath, compressed);
+  static writeBlock(chainPath, block) {
+    const blockPath = this.getBlockPath(chainPath, block.number);
+    if (!fs.existsSync(blockPath)) {
```
Comment (platfowner) [inline blockchain/block-file-util.js:73]: Would it be helpful to log some message when it already exists?

File: blockchain/block-file-util.js (hunk 1 above)
Comment (cshcomcom) [reply blockchain/block-file-util.js]: Done! Thanks

File: common/constants.js (hunk 2 above)
Comment (cshcomcom) [reply common/constants.js]: Done!

File: blockchain/block-file-util.js (hunk 3 above)
Comment (cshcomcom) [reply blockchain/block-file-util.js:55]: Thanks! Done

File: blockchain/block-file-util.js (hunk 4 above)
Comment (cshcomcom) [reply blockchain/block-file-util.js]: Looks good! I changed it to `!blockHash || ChainUtil.isNumber(blockNumber) || blockNumber < 0)`! Done

File: blockchain/block-file-util.js (hunk 5 above)
Comment (cshcomcom) [reply blockchain/block-file-util.js:82]: Done!

**Comment by cshcomcom** [comment]
All comments are addressed! PTAL @platfowner @liayoo

File: blockchain/block-file-util.js [hunk 6]
```diff
@@ -3,75 +3,106 @@ const glob = require('glob');
...
+      const compressed = zlib.gzipSync(Buffer.from(JSON.stringify(block)));
+      fs.writeFileSync(blockPath, compressed);
+    } else {
+      logger.info(`${blockPath} file already exists!`);
+    }
+  }
+
+  static writeHashToNumber(chainPath, blockHash, blockNumber) {
+    if (!blockHash || ChainUtil.isNumber(blockNumber) || blockNumber < 0) {
```
Comment (platfowner) [inline blockchain/block-file-util.js:91]: Did you mean `!ChainUtil.isNumber()`?

Comment (cshcomcom) [reply blockchain/block-file-util.js]: Right! I fix it and run test_integration

**Review by platfowner — APPROVED** [review]
LGTM
Thanks!

Before merging could you try to do some simple experiments?
- Restart a node with block files untouched
- Restart a node with block files removed

**Comment by cshcomcom** [comment]
Thanks I'll try it!

**Review by liayoo — APPROVED** [review]
LGTM, thanks!! 👍

**Comment by cshcomcom** [comment]
Thanks for reviews! I tested 2 case (Restart a node with block files untouched, Restart a node with block files removed.

**Pull request merged**